Refine Domain.{at_exit,at_startup,at_first_spawn,at_each_spawn} callback semantics

kayceesrk opened the pull request:
This PR addresses issues pointed out in #11176 and #11178. It makes the following changes:

* `Domain.at_exit` is now domain-local.
* `Domain.at_exit` also used to ignore exceptions. Now, these exceptions are surfaced as the result of domain execution (similar to how `Stdlib.at_exit` works).
* `Domain.at_startup` is renamed to `Domain.at_each_spawn`.
* `Domain.at_first_spawn` is renamed to `Domain.before_first_spawn`.
* `Domain.before_first_spawn` and `Domain.at_each_spawn` run the callbacks in FIFO order (as requested in #11178). 

The PR also adds tests to the test suite.

xavierleroy commented on stdlib/domain.ml:
```diff
@@ -151,31 +163,35 @@ let do_at_first_spawn () =
-    Stdlib.at_exit wrapped_f
-  else at_exit f
+let at_exit f =
+  if is_main_domain () then
+    Stdlib.at_exit f
+  else begin
+    let old_exit : unit -> unit = DLS.get at_exit_key in
+    let new_exit () = f (); old_exit () in
+    DLS.set at_exit_key new_exit
+  end
```
I wasn't expecting this special case for the main domain.  If a domain other than the main domain calls `Stdlib.exit`, it's going to run exit handlers not associated with itself...  

An alternative could be for `Stdlib.exit` to run the at-exit handlers for the calling domain (those registered with `Domain.at_exit`), then the process-wide handlers registered with `Stdlib.at_exit`.

kayceesrk replied:
I wasn't paying attention to `Stdlib.exit`. With https://github.com/ocaml/ocaml/pull/11213/commits/86627bf875e387194f704d94ce0829e9b627f258, `Stdlib.exit` calls domain-local callbacks and then the process-wide callbacks registered with `Stdlib.at_exit`. This also removes the special casing for main domain in `Domain.at_exit`.

I discovered a separate issue with exceptions raised by `at_exit` callbacks and `Stdlib.exit`. I've made an issue for this: https://github.com/ocaml/ocaml/issues/11221.

kayceesrk replied:
Let me know what you think about the latest commit https://github.com/ocaml/ocaml/commit/86627bf875e387194f704d94ce0829e9b627f258.

xavierleroy reviewed (approved):
Globally this is a big improvement over the current `Domain.at_exit` API, so I'm in favor.  I'm not sure I agree with the semantics of `Domain.at_exit` for the main domain, though; see below.

kayceesrk commented:
@gasche what do you think? Asking since you'd reported #11176.

xavierleroy reviewed (approved):
Looks good to me, but another review or two would be welcome, because it's a stdlib change, and because it's an API that is hard to get right.

kayceesrk commented:
@gasche ping :-)

gasche commented on stdlib/domain.ml:
```diff
@@ -139,7 +151,7 @@ let at_first_spawn f =
   else begin
     let old_f = !first_spawn_function in
-    let new_f () = f (); old_f () in
+    let new_f () = old_f (); f () in
```
Here it is important (and not obvious) that we are running domain-creation callbacks in FIFO order to be symmetric with the domain-exit callbacks in LIFO order. Could you write a comment that makes this intentional design choice explicit in the code?

kayceesrk replied:
This makes sense, but I believe that the comment that you propose should be attached to `at_each_spawn` and `at_exit`. There's no dual for `at_first_spawn`. There is no strong reason for `at_first_spawn` to be FIFO apart from the fact that `at_each_spawn` is FIFO and choosing LIFO for `at_first_spawn` would be more confusing.

gasche replied:
The new comment on `at_exit` is okay with me. Thanks!

gasche commented on stdlib/domain.ml:
```diff
@@ -187,19 +201,20 @@ let spawn f =
       | x -> Ok x
       | exception ex -> Error ex
     in
-    do_at_exit ();
```
If I read the code correctly, now `do_at_exit ()` is only called when the domain callback `f` terminates successfully, but before it would be called even when `f` fails with an exception. Is this correct? Is this change intentional? Is the behavior documented somewhere?

kayceesrk replied:
Good catch. The intention was captured in the PR message:

> `Domain.at_exit` also used to ignore exceptions. Now, these exceptions are surfaced as the result of domain execution (similar to how `Stdlib.at_exit` works).

I messed up the execution. 

My intention was to match how `Stdlib.at_exit` works. Here's what I shall go for in the fix.

----

If the program terminates normally, and `at_exit` raises an exception, the exception is reported. 

```bash
$ cat exn_exit1.ml 
exception E
exception F

let _ =
  at_exit (fun _ ->
    print_endline "in at_exit callback";
    raise E)
$ ocamlopt -g -o exn_exit1.exe exn_exit1.ml
$ ./exn_exit1.exe 
in at_exit callback
Fatal error: exception Exn_exit1.E
```

----

If the program raises an exception, then the `at_exit` callbacks are run. If the `at_exit` callbacks also raise an exception, then the exception is ignored, and the original exception from the program is reported. 

```bash
$ cat exn_exit2.ml
exception E
exception F

let _ =
  at_exit (fun _ ->
    print_endline "in at_exit callbacks";
    raise E);
  raise F
$ ocamlopt -g -o exn_exit2.exe exn_exit2.ml
$ ./exn_exit2.exe 
in at_exit callback
Fatal error: exception Exn_exit2.F
```

The exceptional behaviour of `Stdlib.at_exit` callbacks is undocumented and is known to have issues (see https://github.com/ocaml/ocaml/issues/11221). Hence, it is better not to aim to find the right semantics for `Domain.at_exit` now. My aim is to follow the semantics of `Stdlib.at_exit`.

kayceesrk replied:
Done. Please see the changes in the `spawn` function in `domain.ml`.

gasche commented on stdlib/domain.ml:
```diff
@@ -129,6 +129,18 @@ module DLS = struct
+
+let self () = Raw.self ()
+
+external set_name : string -> unit = "caml_ml_domain_set_name"
```
I suppose that this is not needed anymore that #11264 was merged?

kayceesrk replied:
Removed.

gasche commented on stdlib/domain.mli:
```diff
@@ -39,27 +39,25 @@ val self : unit -> id
-    [at_first_spawn] are called in 'last in, first out' order: the function
-    most recently added with [at_first_spawn] is called first.
+(** [at_first_spawn f] calls [f] before the first domain is spawned by the
+    program. The functions registed with [at_first_spawn] are called on the
```
```suggestion
    program. The functions registered with [at_first_spawn] are called on the
```

gasche commented on stdlib/domain.mli:
```diff
@@ -39,27 +39,25 @@ val self : unit -> id
+(** [at_exit f] calls [f] when the current domain exits. Note that [at_exit]
+    callbacks are domain-local and only apply to the calling domain. The
+    registered functions are called in 'last in, first out' order: the function
+    most recently added with [at_exit] is called first. *)
```
Maybe we could comment that it is possible to combine `at_each_spawn` and `at_exit` to register a cleanup function for each future domain.

kayceesrk replied:
Done. Pushed a small example to show how to use `at_each_spawn` and `at_exit` together.

gasche commented on stdlib/domain.mli:
```diff
@@ -39,27 +39,25 @@ val self : unit -> id
 (** [self ()] is the identifier of the currently running domain *)
 
 val at_first_spawn : (unit -> unit) -> unit
```
I wonder if we should consider naming this function `before_first_spawn` instead of `at_first_spawn`, to clarify that it is called on the parent domain, unlike `at_each_spawn` that runs on the child domain.

kayceesrk replied:
`before_first_spawn` is more appropriate. Renamed.

gasche commented on stdlib/domain.mli:
```diff
@@ -39,27 +39,25 @@ val self : unit -> id
+    @raise Invalid_argument if the program has already spawned a domain. *)
+
+val at_each_spawn : (unit -> unit) -> unit
+(** [at_each_spawn f] calls [f] on the newly spawned domains. The function [f]
```
Style/writing comment: I find the first sentence here confusing. Personally I read `[at_each_spawn f] calls [f]` as saying that `f` will be called immediately when this call is evaluated, so the precision "on the newly spawned domains" is weird. (It will be called right away, in parallel on a bunch of domains?) I'm not sure what formulation to suggest.

This comment somewhat applies to the documentation of the three functions that invariably starts with `[foo f] calls [f] ...`. Maybe a consistently different opening could be used. At the same time the previous implementation-focused style (`[foo f] registers [f] to be called when ...`) was maybe a bit too low-level. Do people have recommendations?

Note: the doc of the other two functions start as follows:
- `[at_exit f] calls [f] when the current domain exits.`
- `at_first_spawn f] calls [f] before the first domain is spawned by the program.`

Somehow I find them slightly less confusing; in particular`calls [f] when ...` is less confusing thanks to the "when" that makes it clear that `calls [f]` does not mean "right away". Still I would prefer a clearer formulation, hopefully that would work in the three cases.

kayceesrk replied:
I'm going with `[foo f] registers [f] to be called`. This wording feels least likely to be misinterpreted.

kayceesrk commented:
@gasche I've addressed all of your comments.

gasche commented on stdlib/domain.ml:
```diff
@@ -129,56 +129,80 @@ module DLS = struct
+
+let self () = Raw.self ()
+
+let is_main_domain () = (self () :> int) == 0
```
Nitpick: for me `== 0` is a code smell / C-ism, I would expect `= 0` instead.

kayceesrk replied:
Fixed.

gasche commented on stdlib/domain.ml:
```diff
@@ -129,56 +129,80 @@ module DLS = struct
     !first_spawn_function();
     (* Release the old function *)
     first_spawn_function := (fun () -> ())
   end
```
For the record: I wondered whether it would be possible to remove specific runtime support for `before_first_spawn`, and implement it instead as a function that gets called by `at_each_spawn`, and maintains mutable state to actually run callbacks only once. But in fact this is not easily possible:

- Running in FIFO order means that we cannot just install a first callback that sets `Atomic.set already_run true`, and have other at_first_spawn callbacks check that. Something more complex would be necessary.
- The fact that `before_first_spawn` runs on the parent domain while `at_each_spawn` runs on the children domain against is a mismatch that I don't know how to solve.

kayceesrk replied:
Thanks for the summary. The fact that `before_first_spawn` runs on the parent domain before any new domains are spawned makes reasoning about concurrency easier.

favonia replied:
Out of curiosity---why do we need `before_first_spawn`?

gasche replied:
It's used in the Format implementation to dynamically switch some functions from a fast sequential implementation to a slower concurrent versions. Other users might way to use the same trick when porting their code to OCaml 5, to preserve the performance profile for single-domain programs.

gasche commented on stdlib/domain.ml:
```diff
@@ -187,25 +211,45 @@ let spawn f =
+                 [at_exit] callbacks and return the original exception. *)
+              result
+          end
+    in
```
This is a lot of code for not-clearly-desirable behavior, but I don't know of a simple way to do much shorter, and the idea of following `at_exit` is right.

kayceesrk replied:
Indeed. Hopefully, we can clear this up when we find a neat solution for unhandled exceptions: https://github.com/ocaml/ocaml/issues/11074.

gasche reviewed (commented):
Thanks! I like the new documentation wording.

kayceesrk commented:
@gasche Once you are happy with the changes, I can squash the commits.

gasche reviewed (approved):
Here is the second approval this PR needed. (In retrospect I agree with @xavierleroy that having two people look was nice.)

I'll let @kayceesrk decide to act on the nitpick or not, and then cleanup the history, and then this is good for merging.

gasche commented:
(Looks like I had missed some recent messages, oh well.)

kayceesrk committed:
Refine the semantics of domain callbacks.

This PR (#11213) addresses issues pointed out in #11176 and #11178. It
makes the following changes:

* `Domain.at_exit` is now domain-local.
* `Domain.at_exit` also used to ignore exceptions. Now, these exceptions
  are surfaced as the result of domain execution (similar to how
  `Stdlib.at_exit` works).
* `Domain.at_startup` is renamed to `Domain.at_each_spawn`.
* `Domain.at_first_spawn` is renamed to `Domain.before_first_spawn`.
* `Domain.before_first_spawn` and `Domain.at_each_spawn` run the
  callbacks in FIFO order (as requested in #11178).

The PR also adds tests to the test suite.

kayceesrk commented:
Thanks for the reviews. I've squashed the commits and have cleaned up the commit message. The CI is also green now. Merging this.
